### PR TITLE
Address set_xlog/ylog problems with DataPHA/Data1DInt classes (#981)

### LINKED ISSUE #981: set_[xy]log/linear do not work for Data1DInt and DataPHA data
```
In [1]: from sherpa.astro.ui import *
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: load_arrays(1, [10, 70, 300], [1, 2,3])

In [3]: load_arrays(2, [10, 70, 300], [50, 100, 500], [1, 2,3], Data1DInt)

In [4]: load_arrays(3, [1, 2, 3], [2, 3, 4], DataPHA)

In [5]: set_xlog()

In [6]: plot('data', 1, 'data', 2, 'data', 3, linestyle='-')
```

FIG

Note that only the top-left plot has a log scale for the x axis.

The problem (I think) is that the set_xxx routines do not know about the new data objects for Data1DInt/DataPHA plots added in #906 and #931. I imagine we need to have a version in both sherpa.ui.utils and sherpa.astro.ui.utils (as only the latter has the PHA models).

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Fix the set_xlog/ylog routines for PHA and 1D integrated datasets. Fixes #981

# Details

The first commit adds tests that show issue #981. They require matplotlib (since you only see the problem when
you create the plot, not with the data returned by get_data_plot). Tests that fail are marked as xfail. Unfortunately, thanks to the
way the tests are structured, several tests are marked as xfail even though they pass, but I think that's okay here.

The second commit just moves to a "more Pythonic" approach (just access the attributes we need, handling any possible error, rather than look for a particular class hierarchy).

The third commit handles the set_x/y[log|linear] routine for DataPHA objects. It also fixes a subtle bug introduced in #931 where use of the clean method would not have reset the plot options for DataPA objects (a test has been added to check this fix).

The fourth commit changes the tests since they can be run without requiring matplotlib. This requires a little bit of work as we now have to query a preference setting rather than ask the axis what the scaling is, which ends up requiring more knowledge of the internal data structures.

The fifth commit just extends the number of plot types that get tested.

The sixth commit fixes the problem for the Data1DInt data type, by adding entries to the _plot_types dictionary. This only fixes the case when the 'all' option (the default) is used: it wouldn't work for 'set_xlog("data")' for Data1DInt datasets.

The seventh commit adds more tests. In particular it adds tests for the 'set_xlog("data")' command (which are marked as xfail).

The eighth commit fixes the set_xlog/ylog calls for DataPHA classes and also addresses the set_xlog('data') and set_ylog('model') style options for all data types by changing the _plot_types dictionary so the values are a list of plot types rather than a single plot type. This means we can identify all the data types as "data" rather than having data, astrodata, data1dint, ... It means we can remove the _set_plot_type method in sherpa.astro.ui.Session since it can now just use the parent classes method.

The last commit are typos and noting an apparently-odd construct that Warren noted in the tests.

### CODE REVIEW

File: sherpa/astro/ui/tests/test_astro_ui_plot.py [hunk 1]
```diff
@@ -2954,3 +2956,609 @@ def test_pha1_plot_multiple_args(clean_astro_ui, basic_pha1):
...
+                          ('fit', True),
+                          ('resid', False),
+                          ('ratio', False),
+                          ('delchi', False)
+                         ])
+def test_set_plot_opt_y(cls, datafunc, plotfunc, answer):
+    """Does set_ylog/ylinear work?
+
+    Unlike test_set_plot_opt_x, the Y axis setting of the plot
+    does not necessarily folloe the set_ylog setting (e.g.
```
Comment (wmclaugh) [inline sherpa/astro/ui/tests/test_astro_ui_plot.py:3135]: folloe -> follow

File: sherpa/astro/ui/tests/test_astro_ui_plot.py [hunk 2]
```diff
@@ -2954,3 +2956,609 @@ def test_pha1_plot_multiple_args(clean_astro_ui, basic_pha1):
...
+    # Create the plot
+    plot()
+    p1 = pdata()
+    if plotfunc in ['fit', 'bkg_fit']:
+        assert p1.dataplot.histo_prefs['xlog']
+        # Check the current behavior of the model plot in case it changes.
+        # Note that for DataPHA datasets the modelplot object is
+        # created on-the-fly using sherpa.astro.plot.ModelPHAHistogram,
+        # rather than a session._plotobj value that is also set in
+        # session._plot_types, so it doesn't get changed bu set_xlog etc.
```
Comment (wmclaugh) [inline sherpa/astro/ui/tests/test_astro_ui_plot.py:3250]: bu -> by

File: sherpa/astro/ui/tests/test_astro_ui_plot.py [hunk 3]
```diff
@@ -2954,3 +2956,609 @@ def test_pha1_plot_multiple_args(clean_astro_ui, basic_pha1):
...
+    plot()
+    p1 = pdata()
+    if plotfunc == 'fit':
+        if is_int:
+            assert p1.dataplot.histo_prefs['ylog'] == answer
+            assert p1.modelplot.histo_prefs['ylog'] == answer
+        else:
+            assert p1.dataplot.plot_prefs['ylog'] == answer
+            assert p1.modelplot.plot_prefs['ylog'] == answer
+    elif plotfunc in ['resid', 'ratio', 'delchi']:
```
Comment (wmclaugh) [inline sherpa/astro/ui/tests/test_astro_ui_plot.py:3168]: Will the assertions ever change here? If not, would it be more efficient to remove the 
 
    elif plotfunc in ['resid', 'ratio', 'delchi']:
        assert p1.plot_prefs['ylog'] == answer

since the same code is executed in the default/fallthrough case.

File: sherpa/astro/ui/tests/test_astro_ui_plot.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_plot.py]: Fixed

File: sherpa/astro/ui/tests/test_astro_ui_plot.py (hunk 2 above)
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_plot.py]: Fixed.

File: sherpa/astro/ui/tests/test_astro_ui_plot.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_plot.py:3168]: It actually isn't the same as the fall through case. since this check needs to be run before the is_int check (ie if you take it out the test fails). I've added a comment to note this.

Comment (DougBurke) [reply sherpa/astro/ui/tests/test_astro_ui_plot.py:3168]: I perhaps should have added it to other tests that use the same ordering, but there's unfortunately a lot of similar-but-not-quite-similar-enough code that I felt it worth abstracting over (since I've found that just gets really hard to follow).

**Comment by DougBurke** [comment]
Sorry for the delay but I was out at the Doctor's this morning

**Comment by anetasie** [comment]
I tested the code in the branch and the bug is fixed. I've tried all the plot_ options with 1 and 2 data sets. Here is part of the relevant history of the session using version:    `4.12.2+9.g6c31af1a` on MacOS with Python 3.8

```
plot_fit(2)
notice(None,0.5)
notice(7.0, None)
fit()
plot_bkg()
plot_bkg(2)
notice(7.0, None)
plot_data(2)
notice()
plot_data()
plot_data(2)
notice(0.5,7.)
plot_data()
plot_data(2)
notice()
notice_id(2,0.5,7.)
plot_fit(2)
notice_id(2)
plot_fit(2)
notice_id(2,0.5,7.)
plot_fit(2)
set_xlinear()
plot_fit(2)
plot_fit(1)
ignore(None,0.5)
ignore(7.0,None)
fit()
plot('fit 1','fit 2')
plot('fit1','fit2')
plot("fit", 1, "fit", 2)
set_xlinear()
plot("fit", 1, "fit", 2)
set_xlog()
plot("fit", 1, "fit", 2)
plot("data", 1, "data", 2)
plot_model()
plot_source()
plot_arf()
set_ylog()
plot_arf()
set_ylinear()
plot_arf()
plot_data()
plot_data(2)
plot("data", 1, "data", 2, ylog=True)
plot_data(2)
plot("data", 1, "data", 2, xlog=False, ylog=True)
plot_data()
set_ylog()
plot_data()
plot("data", 1, "data", 2, xlog=False, ylog=True)
plot_data()
```

**Comment by DougBurke** [comment]
@anetasie - just to confirm, you were using a PHA dataset for this set of plots?

**Comment by anetasie** [comment]
@DougBurke Yes, I have tested the PHA data sets only. I did not try the other types of data.

**Review by dtnguyen2 — COMMENTED** [review]
Is anyone seeing this?:
```
(base) [dtn@devel12 sherpa]$ git pull https://github.com/DougBurke/sherpa.git fix-981
From https://github.com/DougBurke/sherpa
 * branch              fix-981    -> FETCH_HEAD
Already up to date.
(base) [dtn@devel12 sherpa]$ python setup.py develop
  File "setup.py", line 38
    ), file=sys.stderr)
           ^
SyntaxError: invalid syntax
(base) [dtn@devel12 sherpa]$ 
```

Oops, I forgot to type:
```
source activate sherpa
```
before compiling

**Review by dtnguyen2 — APPROVED** [review]

**Comment by DougBurke** [comment]
Ta @dtnguyen2
